### DISCUSSION TITLE
Feature Request: Add newlines when editing multiline history in CLI, improve shortcuts

### DISCUSSION THREAD

**Discussion by daviewales** [duckdb/duckdb#12594] [Ideas]
## Newlines

Sometimes when editing CLI history for a multiline SQL command, it would be nice to insert a newline.

For example, suppose I have the following history:

``` sql
select * from table1
where id = 2
```

and I want to add a join to `table2`, such that I have the following SQL:

``` sql
select * from table1
inner join table2
    on table1.id = table2.id
where id = 2
```

Currently, the only way I can do this is to copy the first line, type the two new lines, then copy the final line.
If I put my cursor at the end of line one and press 'Enter', then the cursor inserts a newline after the last line, not after the current cursor position.

I'm imagining that if there is a semicolon at the end of the last line, then pressing 'Enter' at any point would execute the query, but if the semicolon is missing, it indicates that I haven't finished editing, and I want to be able to insert newlines at any position when I press 'Enter'. (The other option would be to always assume I want a newline when I press 'Enter', and only submit the query if I press 'Enter' when my cursor is on the final character of the final line after a semicolon.) The `IPython` prompt probably comes closest to what I'm looking for here.

## Shortcuts

Sometimes I want to insert text at the start or the end of a line.
Normally I would use the `Home` or `End` keys to jump me to the start or the end of the line.
However, in the DuckDB shell, pressing `Home` jumps me to the first character on the first line, and pressing `End` jumps me to the last character on the last line.

**Comment by rmx90210** [comment]
+1 that it would be super helpful to be able to `Shift`+`Enter` to insert a newline, for example! Thank you for hearing our feedback and considering this, DuckDB team 😄

**Comment by sprckt** [comment]
Yes please, this would be really useful for evolving queries during data exploration

**Comment by l1x** [comment]
Is there a blocker for this? Is shift+enter used for something else

**Comment by hynkle** [comment]
<kbd>ctrl</kbd> + <kbd>x</kbd> can be used to insert a newline at the cursor. (found [here](https://duckdb.org/docs/stable/clients/cli/editing.html))

**Comment by Mytherin** [comment]
Unfortunately the shift modifier cannot be recognized in regular terminals, as such Shift + Enter is not usable as a keyboard shortcut. As mentioned, you can use Ctrl+X to insert a newline (see https://github.com/duckdb/duckdb/pull/10527).

**Reply by daviewales** [reply]
Thanks. Any thoughts about Home and End taking you to the start and end of lines, rather than the start and end of the block?

**Reply by Mytherin** [reply]
After https://github.com/duckdb/duckdb/pull/18490 (which will land in v1.4.0) Ctrl+A and Ctrl+E can be used to return to start/end of line
